### ISSUE TITLE
Several clients run into `basic.publish` when transient flow control kicks in

### ISSUE THREAD

**Issue by michaelklishin** [rabbitmq/rabbitmq-server#156]
See this [mailing list thread](https://groups.google.com/forum/#!topic/rabbitmq-users/qrzFwPd8jms). Relevant parts are captured [in a gist](https://gist.github.com/michaelklishin/fe20cbc524b750391a04).

**Comment by michaelklishin** [comment]
So far I could not reproduce this but a user could with 2 clients (.NET and Python). So there may be something broken.

**Comment by michaelklishin** [comment]
[libpcap capture](https://drive.google.com/file/d/0B12DYRM8lqNoeUVmU0R1dnFYaFk/view?usp=sharing) (abridged) provided.

**Comment by michaelklishin** [comment]
Note: this is when the server is running on Windows (as is the publisher).

**Comment by michaelklishin** [comment]
The OP has started [another list thread](https://groups.google.com/forum/#!topic/rabbitmq-users/56JDH6Di04s). Not much new information there but should be pointed out.

**Comment by rjrizzuto** [comment]
Using RabbitMQ broker version 3.4.4 with my .Net app (using RabbitMQ.Client 3.4.3) I was able to run for about 1/2 an hour, and publish 46GB of data (24 million messages) to the queue using persistent messages.  The publisher did eventually crash, but there is no unexpected_frame message in the broker log. The last item in the broker log is:

=INFO REPORT==== 15-May-2015::10:35:29 ===
vm_memory_high_watermark set. Memory used:3419764032 allowed:3405098188

=WARNING REPORT==== 15-May-2015::10:35:29 ===
memory resource limit alarm set on node rabbit@CYAN199B.

**********************************************************
*** Publishers will be blocked until this alarm clears ***
**********************************************************

The erl.exe process is at 3.4G working set.  The connection shows as blocked in the web page, the channel shows as idle.

The exception the publisher logged is:

Unhandled Exception: System.IO.IOException: Unable to write data to the transport connection: A connection attempt failed because the connected party did not properly respond after a period of time, or established connection failed because
s.SocketException: A connection attempt failed because the connected party did not properly respond after a period of time, or established connection failed because connected host has failed to respond
   at System.Net.Sockets.Socket.Send(Byte[] buffer, Int32 offset, Int32 size, SocketFlags socketFlags)
   at System.Net.Sockets.NetworkStream.Write(Byte[] buffer, Int32 offset, Int32 size)
   --- End of inner exception stack trace ---
   at System.Net.Sockets.NetworkStream.Write(Byte[] buffer, Int32 offset, Int32 size)
   at System.IO.BufferedStream.Flush()
   at System.IO.BinaryWriter.Flush()
   at RabbitMQ.Client.Impl.SocketFrameHandler.WriteFrame(Frame frame)
   at RabbitMQ.Client.Impl.Command.Transmit(Int32 channelNumber, Connection connection)
   at RabbitMQ.Client.Impl.SessionBase.Transmit(Command cmd)
   at RabbitMQ.Client.Impl.ModelBase.ModelSend(MethodBase method, ContentHeaderBase header, Byte[] body)
   at RabbitMQ.Client.Framing.Impl.Model._Private_BasicPublish(String exchange, String routingKey, Boolean mandatory, Boolean immediate, IBasicProperties basicProperties, Byte[] body)
   at RabbitMQ.Client.Impl.ModelBase.BasicPublish(String exchange, String routingKey, Boolean mandatory, Boolean immediate, IBasicProperties basicProperties, Byte[] body)
   at RabbitMQ.Client.Impl.ModelBase.BasicPublish(String exchange, String routingKey, IBasicProperties basicProperties, Byte[] body)
   at rabbitPublisher.rabbitPublisher.Main(String[] args) in c:\Users\rizzuto\Documents\Visual Studio 2012\Projects\queuing\rabbitPublisher\rabbitPublisher.cs:line 55

My guess is that the memory is used up due to the metadata/indexes that the broker keeps in memory, causing the connection to block and the publisher to die.  If so, this may just be expected behavior, but I wanted to provide all the details in case they are pertinent.

**Comment by michaelklishin** [comment]
Yes, eventually a blocked publisher will see a socket write timeout exception, this is expected.

**Comment by rjrizzuto** [comment]
Great.  It looks like I might be able to handle that in code by using the Connection.Blocked and Unblocked event handlers, and not publishing while blocked.

In any event, I'll use 3.4.4 for my testing for now.  

Out of curiosity, does 3.5.2 use as much memory for indexes?

**Comment by michaelklishin** [comment]
@rjrizzuto in a word: yes.

**Comment by rjrizzuto** [comment]
Results from rabbitmqctl status seem to confirm that:

Status of node rabbit@CYAN199B ...
[{pid,8552},
 {running_applications,
     [{rabbitmq_management,"RabbitMQ Management Console","3.4.4"},
      {rabbitmq_web_dispatch,"RabbitMQ Web Dispatcher","3.4.4"},
      {webmachine,"webmachine","1.10.3-rmq3.4.4-gite9359c7"},
      {mochiweb,"MochiMedia Web Server","2.7.0-rmq3.4.4-git680dba8"},
      {rabbitmq_management_agent,"RabbitMQ Management Agent","3.4.4"},
      {rabbit,"RabbitMQ","3.4.4"},
      {mnesia,"MNESIA  CXC 138 12","4.12.5"},
      {os_mon,"CPO  CXC 138 46","2.3.1"},
      {sasl,"SASL  CXC 138 11","2.4.1"},
      {inets,"INETS  CXC 138 49","5.10.6"},
      {amqp_client,"RabbitMQ AMQP Client","3.4.4"},
      {xmerl,"XML parser","1.3.7"},
      {stdlib,"ERTS  CXC 138 10","2.4"},
      {kernel,"ERTS  CXC 138 10","3.2"}]},
 {os,{win32,nt}},
 {erlang_version,
     "Erlang/OTP 17 [erts-6.4] [64-bit] [smp:8:8] [async-threads:30]\n"},
 {memory,
     [{total,3411540720},
      {connection_readers,9064},
      {connection_writers,2704},
      {connection_channels,9104},
      {connection_other,28024},
      {queue_procs,6671128},
      {queue_slave_procs,0},
      {plugins,511352},
      {other_proc,13991432},
      {mnesia,65512},
      {mgmt_db,184856},
      {msg_index,2935088024},
      {other_ets,415194904},
      {binary,14415344},
      {code,19975441},
      {atom,703377},
      {other_system,4690454}]},
 {alarms,[memory]},
 {listeners,[{clustering,25672,"::"},{amqp,5672,"127.0.0.1"}]},
 {vm_memory_high_watermark,0.4},
 {vm_memory_limit,3405098188},
 {disk_free_limit,50000000},
 {disk_free,806677446656},
 {file_descriptors,
     [{total_limit,8092},
      {total_used,6},
      {sockets_limit,7280},
      {sockets_used,2}]},
 {processes,[{limit,1048576},{used,196}]},
 {run_queue,0},
 {uptime,6686}]

**Comment by rjrizzuto** [comment]
Have you had any success reproducing this at least?  It seems it is sensitive to the amount of data published to the queue, not the speed, since the python script I wrote published at ~300 messages/sec vs 9000 for .net.

**Comment by dumbbell** [comment]
Hi!

I reproduced the issue with @rjrizzuto's Python test script, twice in a row. The Python catch an exception because the socket is closed by RabbitMQ. I captured the network packets and need to review them. But as the bug happens with both the .NET client and Pika, it must lie somewhere in the broker.

In RabbitMQ's logs:
```
=ERROR REPORT==== 1-Jun-2015::12:59:25 ===
Error on AMQP connection <0.1161.0> ([::1]:48981 -> [::1]:5672, vhost: '/', user: 'guest', state: running), channel 1:
{amqp_error,unexpected_frame,
            "expected method frame, got non method frame instead",none}

=INFO REPORT==== 1-Jun-2015::12:59:25 ===
closing AMQP connection <0.1161.0> ([::1]:48981 -> [::1]:5672)
```

I limited RabbitMQ to ~500 MiB of memory by setting `vm_memory_high_watermark`. The bug happens when the queue reaches `vm_memory_high_watermark_paging_ratio` (~250 MiB) and it needs to be paged out. 

One thing we changed in 3.5.3 is how we clear the read cache when we need to reclaim memory. I don't think this is the bug. However, I suspect the problem is related to our read buffers handling and this change increases the chance to hit it. That would explain why you have it with RabbitMQ 3.4.x but less often.

**Comment by michaelklishin** [comment]
@dumbbell I'm inclined to think it's something in the reader state machine as I believe it can be reproduced (not as easily but still) with earlier versions. We've had reader changes in 3.5.0, too, to avoid logging connections that sent no data (e.g. load balancer aliveness checks).

**Comment by rjrizzuto** [comment]
I'm glad to hear that you were able to reproduce it.  Hopefully the fact that it is more reproducible in 3.5.x will make it easier to fix.

**Comment by dumbbell** [comment]
After more investigation, it is not that "simple".

In all the network captures I did (using Pika as the client on the same host as RabbitMQ), I see the following scenario:

1. Pika sends perfect sequences of `Basic.Publish -> Content-Header -> Content-Body`, one AMQP frame per TCP packet.
2. At some point, Pika sends 3 or 4 TCP packets filled to reach the MTU (in my case, 16384). Those packets obviously contain multiple perfect sequences.
3. Then Pika goes back to one AMQP frame per TCP packet. However, this packet contains an unexpected frame. For instance, the last "big" TCP packet finished with a `Content-Body`, then the small TCP packet contains a `Content-Header` or a second `Content-Body`. In all cases, the `Basic.Publish` frame is missing.

I think it happens at a time where RabbitMQ does not read the network (the reader is blocked because the memory is under pressure), but I can't confirm if this is really linked. So far, 66% of the captures shows that a `Connection.Unblocked` was sent by RabbitMQ to the client just before the big TCP packets happen.

I will look at Pika source code to see how it works.

**Comment by michaelklishin** [comment]
While other processes and back pressure can affect the reader, the state machine should not enter this incorrect state on valid input. So that may be a red herring.

**Comment by rjrizzuto** [comment]
@dumbbell Keep in mind the same symptoms show up at the server with either a Python or C# publisher.  Unless they copied from each other, it seems more likely that the issue is in common code, i.e. the RabbitMQ server component.

I was also trying to look at the packet capture with wireshark, but having some difficulty, so I am not sure how much to trust it.

One question I had was how the client libraries handle a connection that is in flow control.  Does a tcp send return e_wouldblock or the like, and does the client lib retry later, or skip that send and go on to the next?

**Comment by michaelklishin** [comment]
They either block on `write(2)` or catch `EWOULDBLOCK` and re-try.

**Comment by dumbbell** [comment]
I compared the captures made by Wireshark to the data processed in `rabbit_reader.erl` and they both match. So the frames received by RabbitMQ are incorrect and my test case doesn't demonstrate a problem in RabbitMQ.

Maybe the issue exposed by the .NET client library is different.

**Comment by dumbbell** [comment]
As a reference, here is the Python script:
```python
#!/usr/bin/env python
import pika
import sys
import time

connection = pika.BlockingConnection(pika.ConnectionParameters(host='localhost'))
channel = connection.channel()
channel.queue_declare('test', durable=True)

message =  b'\x00' * 2048
while 1:
    channel.basic_publish(exchange='', routing_key='test', body=message)
```

**Comment by dumbbell** [comment]
So, I'm using Pika 0.9.14 (the latest release) and after looking at what it is doing, I confirm the problem comes from Pika in my case.

1. Frames are written to the wire individually by this function:
 https://github.com/pika/pika/blob/0.9.14/pika/adapters/base_connection.py#L360

 `self.socket.sendall(frame)` either sends the frame successfully or an exception is raised.

2. An exception is raised when RabbitMQ do not accept any more TCP packets. The exception is handled by:
 https://github.com/pika/pika/blob/0.9.14/pika/adapters/blocking_connection.py#L351

 The `if self.is_closing` condition is never met so the exception is ignored (the connection is not closed) and the frame is dropped by Pika.

And because there is no retry and no atomicity in the send of the 3 frames composing `Basic.Publish`, sometimes, the method frame is not sent to RabbitMQ but following frames are, leading the the reported error in RabbitMQ. The code in Pika's `master` branch is completely different so the problem may have been fixed, I did not try it.

So we are back to square 1 with the original issue. I will try to reproduce it with the .NET client.

**Comment by michaelklishin** [comment]
@dumbbell we should let @gmr know and fix 0.9.x as it is the version our tutorials use. Plus this is something we've recommended to the OpenStack community.

**Comment by dumbbell** [comment]
I looked at `master` and I believe pika/pika@cd8c9b08e8a121acfb8b53344292d7b2ff4eecc6 fixes the problem.

**Comment by dumbbell** [comment]
I could reproduce the problem with your `Publisher.cs` test case.

Again, Wireshark and RabbitMQ agree on the received data: the received frames are incorrect. In the last run, RabbitMQ received three `Basic.Publish` frames in a row, without `Content-Header` and `Content-Body` in between.

And again, this happens at a time RabbitMQ's host receive buffer is full (because RabbitMQ blocks the reads). Thus it looks like the problem in Pika.

**Comment by michaelklishin** [comment]
Not just in Pika but in multiple clients then. We should fix it everywhere (by re-trying or even sending all AMQP 0-9-1 frames as a single byte sequence?)

**Comment by rjrizzuto** [comment]
It seems to me that sending Basic.Publish+Content-header+Content-Body in one atomic send may "fix" the issue with the server receiving an invalid sequence, but it would break the implied contract of reliable delivery with the publisher.  Retrying would be necessary to adhere to the contract.

**Comment by michaelklishin** [comment]
@rjrizzuto thank you for your input. The plan after a discussion we had internally is to

 * Send all frames in a single byte array (buffer, string, or similar, depending on the client)
 * Use `select(2)` to check if a write will succeed (or something similar)

The tricky part is telling exceptions that require re-trying vs. those that indicate a dead socket (and thus need to be propagated to start connection recovery).

**Comment by rjrizzuto** [comment]
Thanks for the update.  I'm glad you've got a handle on this.

One thought I had is that it might be helpful to have some logging in the client libraries for exception cases like this.  For .Net, something like Common.Logging would work, and allow the user to decide where to log to.

**Comment by michaelklishin** [comment]
Yes, the .NET client needs logging. I don't know about logging really nitty-gritty details such as this one (ideally we should make things safe regardless, so there's no point in logging it) but suggestions on what we should use for logging are appreciated. Feel free to file an issue for the .NET client!

**Comment by rjrizzuto** [comment]
In general, we try to log exception/error cases only so as not to impact performance.  We do use debug level logging for development purposes, and set the log filtering in production to only include info and higher.

Here's a perfect case in point from AutorecoveringConnection.cs:

                        catch (Exception e)
                        {
                            // TODO: logging
                            Console.WriteLine("BeginAutomaticRecovery() failed: {0}", e);
                        }

With common.logging, I'd do the following:

                        catch (Exception e)
                        {
                            logger.Warn("BeginAutomaticRecovery() failed", e);
                        }

The logger would log all the details, including stack trace (configurable in the logging config file) about the exception.

**Comment by michaelklishin** [comment]
Please move this into a separate issue in rabbitmq/rabbitmq-dotnet-client. We are familiar with logging levels, much less so with what logging library is the most appropriate for a small, 0 dependencies library like our client.

**Comment by rjrizzuto** [comment]
I have moved the logging suggestion.

**Comment by dumbbell** [comment]
I couldn't reproduce the issue with the Java client library and PerfTest, so this one may be good already. I'm going to look at what is done there.

**Comment by dumbbell** [comment]
More tests with more client libraries:
* Pika's `master` branch is fixed.
* Bunny (Ruby) is fine too.

**Comment by michaelklishin** [comment]
Thanks. I'd check amqp.node, Go amqp , and Haskell's Network.AMQP, too.

**Comment by rjrizzuto** [comment]
Have you had any luck finding the issue in the .Net client?  I only tested with Python to see if it was the client lib or the message broker, and apparently that was a red herring anyway.

**Comment by michaelklishin** [comment]
Yes, we understand what the issue is. Will fix it for .NET after s few other .NET issues are merged.

**Comment by rjrizzuto** [comment]
Thanks, that is great news!

**Comment by dumbbell** [comment]
If I understand correctly, with Node.js' amqp.node, handling connection write failure is deferred to the application.

**Comment by dumbbell** [comment]
`streadway/amqp` (Go) is fine.

**Comment by michaelklishin** [comment]
@dumbbell @rjrizzuto I think .NET client is ready for another round (on both stable and master). It now only performs a single flush per `basic.publish` frame set. Please give it a try and let me know if that's sufficient.

**Comment by rjrizzuto** [comment]
If the single write fails, does that mean the data is lost?

If you have a binary, I can give it a try.

**Comment by michaelklishin** [comment]
If a socket write times out, there will be an exception. But it will be obvious what's going on, unlike what's described above.

**Comment by rjrizzuto** [comment]
What I mean is that if the write times out due to flow control, and there is an exception, is the frame set lost, or retried later?  I'm concerned about being in flow control and having the publisher quietly losing data.  In my application, I'd prefer to see the publisher retry or block till the flow control state is lifted.  Of course, other applications may prefer a different policy in such a case.

**Comment by michaelklishin** [comment]
Blocking or retrying is the plan (this is what other clients do).

**Comment by rjrizzuto** [comment]
Thanks for clarifying

**Comment by michaelklishin** [comment]
For those looking to help us test a fix, [here's a build from a recent commit on the stable branch](https://drive.google.com/file/d/0B12DYRM8lqNoV1B3T0pzWFJzSjA/view?usp=sharing).

(if you are paranoid and need a confirmation that indeed it was produced by the RabbitMQ team, contact `mklishin` in Pivotal domain and I'll confirm)

**Comment by michaelklishin** [comment]
I'll close this as we've evaluated the issue and what clients it affects, and if the fixes are not sufficient, new issues should be filed for the affected clients.

**Comment by rjrizzuto** [comment]
I'm out of the office this week, but will give it a try when I get back.

**Comment by rjrizzuto** [comment]
I'm having trouble getting the .Net library.  The company I work at blocks file sharing sites, such as google drive, as well as most email attachments.  It's quite annoying.

**Comment by michaelklishin** [comment]
@rjrizzuto I will email you the .dll. Building from source is also straightforward: there are documents on how to do that in the repo.

**Comment by rjrizzuto** [comment]
You already did, but my company strips binary attachments.

**Comment by rjrizzuto** [comment]
I'm working with our tech department to see if they can get the file from the link you posted.

**Comment by rjrizzuto** [comment]
After running for a while (1.2 million message of size 1024), I get a System.IO.IOException that says "Unable to write data to the transport connection: A connection attempt failed because the connected party did not properly respond after a period of time, or established connection failed because connected host has failed to respond.".  Here's the stack trace:

		StackTrace	"   at System.Net.Sockets.NetworkStream.Write(Byte[] buffer, Int32 offset, Int32 size)\r\n   at System.IO.BufferedStream.Flush()\r\n   at System.IO.BinaryWriter.Flush()\r\n   at RabbitMQ.Client.Impl.SocketFrameHandler.WriteFrameSet(IList`1 frames)\r\n   at RabbitMQ.Client.Framing.Impl.Connection.WriteFrameSet(IList`1 f)\r\n   at RabbitMQ.Client.Impl.Command.TransmitAsFrameSet(Int32 channelNumber, Connection connection)\r\n   at RabbitMQ.Client.Impl.Command.Transmit(Int32 channelNumber, Connection connection)\r\n   at RabbitMQ.Client.Impl.SessionBase.Transmit(Command cmd)\r\n   at RabbitMQ.Client.Impl.ModelBase.ModelSend(MethodBase method, ContentHeaderBase header, Byte[] body)\r\n   at RabbitMQ.Client.Framing.Impl.Model._Private_BasicPublish(String exchange, String routingKey, Boolean mandatory, Boolean immediate, IBasicProperties basicProperties, Byte[] body)\r\n   at RabbitMQ.Client.Impl.ModelBase.BasicPublish(String exchange, String routingKey, Boolean mandatory, Boolean immediate, IBasicProperties basicProperties, Byte[] body)\r\n   at RabbitMQ.Client.Impl.ModelBase.BasicPublish(String exchange, String routingKey, Boolean mandatory, IBasicProperties basicProperties, Byte[] body)\r\n   at RabbitMQ.Client.Impl.ModelBase.BasicPublish(String exchange, String routingKey, IBasicProperties basicProperties, Byte[] body)\r\n   at RabbitMQ.Client.Impl.AutorecoveringModel.BasicPublish(String exchange, String routingKey, IBasicProperties basicProperties, Byte[] body)\r\n   at rabbitPublisher.rabbitPublisher.Main(String[] args) in c:\\Users\\rizzuto\\Documents\\Visual Studio 2012\\Projects\\queuing\\rabbitPublisher\\rabbitPublisher.cs:line 68"	string

Is that expected?  Do I need to handle retry logic in my app for the message that was not sent, or is that done in the library?

I can confirm that the connection is not dropping.

**Comment by rjrizzuto** [comment]
I split out handling of System.IO.IOExceptionto (just print a message) from System.Exception.  I am also getting a System.ObjectDisposedException {"Cannot access a disposed object.\r\nObject name: 'System.Net.Sockets.Socket'."} with this call stack:

		StackTrace	"   at System.Net.Sockets.Socket.Poll(Int32 microSeconds, SelectMode mode)\r\n   at RabbitMQ.Client.Impl.SocketFrameHandler.WriteFrameSet(IList`1 frames)\r\n   at RabbitMQ.Client.Framing.Impl.Connection.WriteFrameSet(IList`1 f)\r\n   at RabbitMQ.Client.Impl.Command.TransmitAsFrameSet(Int32 channelNumber, Connection connection)\r\n   at RabbitMQ.Client.Impl.Command.Transmit(Int32 channelNumber, Connection connection)\r\n   at RabbitMQ.Client.Impl.SessionBase.Transmit(Command cmd)\r\n   at RabbitMQ.Client.Impl.ModelBase.ModelSend(MethodBase method, ContentHeaderBase header, Byte[] body)\r\n   at RabbitMQ.Client.Framing.Impl.Model._Private_BasicPublish(String exchange, String routingKey, Boolean mandatory, Boolean immediate, IBasicProperties basicProperties, Byte[] body)\r\n   at RabbitMQ.Client.Impl.ModelBase.BasicPublish(String exchange, String routingKey, Boolean mandatory, Boolean immediate, IBasicProperties basicProperties, Byte[] body)\r\n   at RabbitMQ.Client.Impl.ModelBase.BasicPublish(String exchange, String routingKey, Boolean mandatory, IBasicProperties basicProperties, Byte[] body)\r\n   at RabbitMQ.Client.Impl.ModelBase.BasicPublish(String exchange, String routingKey, IBasicProperties basicProperties, Byte[] body)\r\n   at RabbitMQ.Client.Impl.AutorecoveringModel.BasicPublish(String exchange, String routingKey, IBasicProperties basicProperties, Byte[] body)\r\n   at rabbitPublisher.rabbitPublisher.Main(String[] args) in c:\\Users\\rizzuto\\Documents\\Visual Studio 2012\\Projects\\queuing\\rabbitPublisher\\rabbitPublisher.cs:line 69"	string

That seems like an issue.

**Comment by michaelklishin** [comment]
I'm going to file a separate issue.

**Comment by michaelklishin** [comment]
https://github.com/rabbitmq/rabbitmq-dotnet-client/issues/108. Unless we can reproduce it quickly it will be for 3.5.5.

**Comment by rjrizzuto** [comment]
I think the System.ObjectDisposedException exception was related to a heartbeat timeout while paused in the debugger.

The System.IO.IOException seems to be the result of flow control, and perhaps that is expected behavior, but if so it may need to be documented, especially for retry scenarios.

**Comment by greggwon** [comment]
We have a single connection to rabbit with both readers and writers using the same connection.  How much contention should we expect in that case.  Also, we seem to see that when the above situation occurs, all readers on a blocked connection seem to stop receiving messages.  Is this expected behavior in this case?  Should we have a single connection per client per queue?

**Comment by michaelklishin** [comment]
@greggwon please post questions to rabbitmq-users or Stack Overflow. RabbitMQ uses GitHub issues for specific actionable items engineers can work on, not questions. Thank you.
